### PR TITLE
Add PTY option for Meterpreter shell command

### PR DESCRIPTION
**Pull request by sn0wfa11**
## Overview
This PR adds a pybash option to the Meterpreter `shell` command to provide a little bit nicer pty shell on NX machines that have python installed. Some Meterpreter flavors (python and php) even produce a color pty depending on target settings.

I like to drop to a shell to manually work on machines and got tired of having to type or paste in the pybash line. Having the option to go straight to this is rather nice. It is not a full tty with tab completion and command history, but it's better than a raw /bin/sh from the standard `shell` command.

Decided to make this an option instead of the default behavior. However, it would be easy to make it the default action, if people like the idea.

### Example Output: (From a Kali target)
```
meterpreter > shell -h
Usage: shell [options]

Opens an interactive native shell.

OPTIONS:

    -b        NX Bash PTY via pybash, requires python
    -h        Help menu.

meterpreter > shell
Process 2398 created.
Channel 3 created.
/bin/sh: 0: can't access tty; job control turned off
# 

meterpreter > shell -b
Process 2404 created.
Channel 4 created.
root@sn0wfa11:/# 
```

### Screenshot: (From a CentOS target)
shell_b

## Verification
- [ ] Start up some Meterpreter shells on various systems:
- [ ] `shell -b` should produce a pty bash on NX machines with python
- [ ] `shell -b` should act just like `shell` on NX machines without python
- [ ] `shell -b` should act just like `shell` on Windows or other non-NX machines

### CODE REVIEW

**Comment by bcoles** [comment]
I like the idea. Not sure if this is the best way to implement, but I don't have a better suggestion.

Golfed for style:


```ruby
  #
  # Spawn a python pty /bin/bash or /bin/sh shell 
  #
  def pybash_shell
    path = '/usr/bin/python'
    return false unless client.fs.file.exist? path

    sh_path = client.fs.file.exist?('/bin/bash') ? '/bin/bash' : '/bin/sh'
    ags = "-c \'import pty;pty.spawn(\"#{sh_path}\")\'"
    cmd_execute('-f', path, '-a', ags, '-c', '-i')

    true
  rescue
    false
  end
```

**Comment by fsacer** [comment]
also keep in mind that you should look for both python2 and python3 executables as it's usual to have only python3 installed

**Comment by fsacer** [comment]
and it would be nice if you could hide the echo back

**Comment by sn0wfa11** [comment]
@bcoles and @fsacer, thank you for the style fixes and the feedback! Updated for style and also for checking more python paths.

@fsacer, I'm still looking into silencing the echo, which is built into the python pty module. I don't want to make this function overly complicated just to hide a command echo. However, if you have any ideas please share!

**Comment by wvu-r7** [comment]
Thanks for the contribution. A few things:

Python Meterpreter calls `pty.openpty()`:

https://github.com/rapid7/metasploit-payloads/blob/b3d4495ef8c8506863b512eb577f9cbb35e1c56c/python/meterpreter/ext_server_stdapi.py#L1032-L1044

```
meterpreter > shell
Process 31256 created.
Channel 1 created.
/bin/sh: 0: can't access tty; job control turned off
$ tty
/dev/pts/5
$
```

```
meterpreter > execute -if bash
Process 32156 created.
Channel 3 created.
bash: cannot set terminal process group (31028): Inappropriate ioctl for device
bash: no job control in this shell
wvu@hiigara:~/metasploit-framework:master$ tty
/dev/pts/7
wvu@hiigara:~/metasploit-framework:master$
```

POSIX Meterpreter used to do it with `ptmx(4)`:

https://github.com/rapid7/metasploit-payloads/blob/6625248fc748dafd714af97780a9872a10faaf87/c/meterpreter/source/extensions/stdapi/server/sys/process/process.c#L94-L147

Ideally this should be added to Mettle instead of a command dispatcher. But this works in a pinch.

`script -q /dev/null` is a more universal solution I've come up with:

```
meterpreter > execute -if script -a "-qc bash /dev/null"
Process 31803 created.
Channel 2 created.
wvu@hiigara:~/metasploit-framework:master$ tty
/dev/pts/7
wvu@hiigara:~/metasploit-framework:master$
```

The syntax may be `script -q /dev/null bash` on other systems, but the command is ubiquitously implemented and calls `openpty(3)` under the covers.

Thoughts? If we can justify that this adds enough over existing functionality, I'll merge this. :)

**Comment by sn0wfa11** [comment]
@wvu-r7, thank you for the feedback and information. Especially the "script" based bash execution! That is new to me and I'm putting that one in the toolkit. Might come in handy as a good restricted shell escape someday too ;) I will adjust the code to use "scrip" either instead of, or in addition to the pybash, and do some testing on various systems before pushing a commit. While I'm working on this, see the information below regarding your questions. Others, please feel free to chime in too. If this is not enough justification to land, let me know, it is not a huge deal and I still learned a few new things.

My goals for the PR were along these lines:
- Provide the option for using a different shell, one that is a little bit easier on the eyes.
- Be able to drop in and out of that different shell easily with one command and an option switch.
- Have the option be consistent across different Meterpreter flavors, python, Mettle, php.

I agree with you that having the execution part in payloads is where it ultimately should reside. It would be great to have this across the different Meterpreters, especially if you could get tab completion and command history. Which it looks like this is a goal, but might be a ways out. Once that stage is reached I'm sure that the shell command in the dispatcher would be revisited anyway. I personally like the idea of having a quick option before those changes arrive, but I'm probably a little bit biased... ;)

**Comment by WangYihang** [comment]
```
socat - exec:'bash -li',pty,stderr,setsid,sigint,sane
```
This is another familiar command to open a shell under pty, just like:
```
python -c 'import pty; pty.spawn("/bin/bash")'
```

**Comment by sn0wfa11** [comment]
Pushed an update that uses multiple options to run bash based on the feedback above. This appears to be working well so far. I will continue to test on different systems. Let me know what you think.

**Comment by sn0wfa11** [comment]
@WangYihang, thanks for the feedback. The reason behind checking full "literal" paths is to avoid as many dependencies as possible.

`which` queries the $PATH environment variable to resolve literal paths. If you only use `python` or `bash` the system queries $PATH to get the literal path before execution. `which` itself also resolves to a literal path.
```
root@kali:~# which which
/usr/bin/which
```
This makes you dependent on $PATH being accessable and correct for the user you are impersonating. Had a CTF box once where $PATH was all messed up. It was a troll, but the lesson stuck with me. Using `which` makes you dependent on that command being present. And `python` is typically a symbolic link to either `python2` or `python3` and may not always be present.

Hope that makes sense.

**Comment by WangYihang** [comment]
@sn0wfa11 It's my fault, I am using `zsh`, and `which` is a builtin command of `zsh`, so I mistakenly thought that which is a builtin command of other shells, thank you for you suggestion~

**Comment by WangYihang** [comment]
I am wondering if there is another command just like which, not looking up binaries in $PATH but search the common binary folders like `/bin` `/sbin` `/usr/bin` ... If there is, I think it will be very helpful...
Or can we just write a shell script to implement the `which` command by searching the common binaries folder?

**Comment by wvu-r7** [comment]
@sn0wfa11, @WangYihang: Consider `expect -c 'spawn /bin/bash; interact'` for an old-school fallback.

**Comment by WangYihang** [comment]
cool, yeah, expect will also open a pseudo-terminal, good idea~

wvu-r7 <notifications@github.com>于2018年8月24日周五 上午3:20写道：

**Comment by wvu-r7** [comment]
It may not be routinely installed on modern systems, but it seems to be there by default on OS X? /shrug

**Comment by sn0wfa11** [comment]
@wvu-r7, I like the `expect` too and I will add it to the list of methods.

I just realized that it would be fairly trivial to add a `-l` option to list the available shells and a `-s <string>` option for spawning a user-defined shell such as /bin/zsh if it is present on the target. I'll work on getting this added and tested. What do you all think?

Also, the current order of methods is below. I'm a concerned about the potential for different syntax with `script`.  I'm inclined to move it down on the list even though it works really well on Linux systems, but the `script -qc /bin/bash /dev/null` syntax fails on Mac's. I'm struggling with finding a good way to check the expected syntax before hand. All the other methods are more consistent. Any thoughts on the order?

Current Method Order:
1. Socat
2. Script
3. Python
4. Expect (To be added)
5. Fall back to standard `/bin/sh`

**Comment by wvu-r7** [comment]
@sn0wfa11: I like to use `||`, but it's a cheap trick:

```
wvu@kharak:~$ echo $SHLVL
1
wvu@kharak:~$ script -qc /bin/bash /dev/null || script -q /dev/null /bin/bash
script: illegal option -- c
usage: script [-adkpqr] [-t time] [file [command ...]]
wvu@kharak:~$ echo $SHLVL
2
wvu@kharak:~$
```

Shellception!

**Comment by wvu-r7** [comment]
@WangYihang: `whereis(1)` will search more than just `$PATH` in certain implementations (maybe just Linux), but its availability and specific functionality are non-standard.

```
wvu@[redacted]:~$ whereis -l | grep ^bin
bin: /usr/bin
bin: /usr/sbin
bin: /usr/lib
bin: /bin
bin: /sbin
bin: /etc
bin: /lib
bin: /lib64
bin: /usr/games
bin: /usr/local/bin
bin: /usr/local/sbin
bin: /usr/local/etc
bin: /usr/local/lib
bin: /usr/local/games
bin: /usr/include
bin: /usr/local
bin: /usr/share
bin: /opt/metasploit-framework/bin
bin: /opt/ircd-hybrid/bin
wvu@[redacted]:~$
```

My fault on those last two, hehe.

OS X:

```
wvu@kharak:~$ whereis -l
whereis: illegal option -- l
usage: whereis program [...]
wvu@kharak:~$ sysctl user.cs_path
user.cs_path: /usr/bin:/bin:/usr/sbin:/sbin
wvu@kharak:~$ whereis rbenv
wvu@kharak:~$ which rbenv
/usr/local/bin/rbenv
wvu@kharak:~$
```

You can probably guess that the BSDs are different, though OS X is usually close to FreeBSD.

Sorry, don't have a great answer. Reinvent the wheel with our own `whereis`? :joy:

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb [hunk 1]
```diff
@@ -40,6 +40,15 @@ class Console::CommandDispatcher::Stdapi::Sys
     "-h" => [ false, "Help menu."						   ],
     "-f" => [ true,  "Force a reboot, valid values [1|2]"			   ])
 
+  #
+  # Options used by the 'shell' command.
+  #
+  @@shell_opts = Rex::Parser::Arguments.new(
+    "-h" => [ false, "Help menu."						   ],
+    "-b" => [ false, "NX Bash Shell."						   ],
```
Comment (wvu-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb]: I'd change references to "NX" to something more descriptive.

**Comment by sn0wfa11** [comment]
@wvu-r7, let me know if those descriptions are better or if you think more needs to be added.

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb [hunk 2]
```diff
@@ -269,10 +317,46 @@ def cmd_shell(*args)
...
+    path = '/usr/bin/socat'   if client.fs.file.exist?('/usr/bin/socat')
+    path = '/usr/bin/script'  if client.fs.file.exist?('/usr/bin/script')  && !path
+    path = '/usr/bin/python'  if client.fs.file.exist?('/usr/bin/python')  && !path
+    path = '/usr/bin/python2' if client.fs.file.exist?('/usr/bin/python2') && !path
+    path = '/usr/bin/python3' if client.fs.file.exist?('/usr/bin/python3') && !path
+    path = '/usr/bin/expect'  if client.fs.file.exist?('/usr/bin/expect')  && !path
+    return false unless path
+
+    # Commands for methods - "export TERM=xterm" provides colors, "clear" command, etc. as available on the target.
+    cmd = "export TERM=xterm; #{path} - exec:'#{sh_path} -li',pty,stderr,setsid,sigint,sane" if path =~ /socat/
```
Comment (wvu-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb]: Consider simply prefixing `TERM=xterm` or using `env(1)` (potentially more portable).

If you require an exported variable in the shell's environment (doesn't look like you do), you should use the portable syntax `TERM=xterm; export TERM`.

http://wiki.bash-hackers.org/scripting/nonportable

**Comment by wvu-r7** [comment]
Thanks, they look great! I'm no longer handling this PR, but I think this about completes my review.

**Comment by sn0wfa11** [comment]
@wvu-r7 -  Thank you so much for your help! I learned a lot working on this PR!

**Comment by wvu-r7** [comment]
No problem. Thanks for the contribution! This is a pretty big step forward for something we've lacked for so long!

**Comment by wvu-r7** [comment]
Since @asoto-r7 is occupied at the moment, and it seems I've been handling this after all, I'll test and land it. Thank you for sticking this through!

**Comment by sn0wfa11** [comment]
Sounds good. Let me know if anything else pops up and needs to be changed. I'll also place a comment on the Feature Request that you made as I would be happy to rework this once PTY support is added to Mettle.

**Comment by wvu-r7** [comment]
@sn0wfa11: I didn't create a PR to your branch because I had to fix some conflicts by merging `master`.

Please see my finalized revision at https://github.com/rapid7/metasploit-framework/compare/master...wvu-r7:pr/10389. There are a number of significant changes. I'm happy to explain or justify any of them.

Note that I consolidated the `-b` and `-s` options into `-t` as one might be familiar with using `ssh -t`.

If you're happy with the diff, we can move forward with this and get it landed. Thank you!

**Comment by PrinceKK300188** [comment]
Kindly just dont do anything and keep my account normal

**Comment by sn0wfa11** [comment]
@wvu-r7, the changes look great! I'm good with getting this thing landed. Thanks again for all the help.

**Comment by wvu-r7** [comment]
# Release Notes

This adds the `-l` and `-t` options to the `shell` command in Meterpreter, providing the ability to list available shells and spawn an arbitrary one (`/bin/bash` by default) with PTY support.

**Comment by q2dg** [comment]
Hello.

I've tried to execute _shell -t_ inside a meterpreter session (corresponding to the linux/x64/meterpreter/reverse_tcp payload) and I can get any interactive shell (as it's seen in image below). Executing _shell_ without arguments is fine as always, though.

shell-t

**Pull request merged**